This is a transference of the ergm-private" "control_llik" branch to the public version ergm 4.0.

handcock opened the pull request:
The branch consolidates control arguments to the likelihood functions used in ergm.estimate into a single list variable called "control.llik".
This simplifies the calls to the functions.

The branch also adds a simple hook in "ergm.estimate" for the Kpenalty likelihood functions. These are required for the "ergm'tapered" package to interoperate with "ergm".

Also fixed a few documentation typos in "ergm".

handcock committed:
This is a transference of the ergm-private" "control_llik" branch to the public version ergm 4.0.

The branch consolidates control arguments to the likelihood functions used in ergm.estimate into a single list variable called "control.llik".
This simplifies the calls to the functions.

The branch also adds a simple hook in "ergm.estimate" for the Kpenalty likelihood functions. These are required for the "ergm'tapered" package to interoperate with "ergm".

Also fixed a few documentation typos in "ergm".

handcock commented:
To make life easy, I created a branch of ergm 4.0 and manually made the edits from the private "control_llik" branch. As it turned out it was necessary as ergm 4.0 had evolved so that automatic merging would have been a mess.

krivit commented:
Thanks, Mark! I'll take a look the week after Sunbelt.

handcock committed:
Added in exclusion for "Taper_Penalty" term from mcmc.diagnostics.

handcock committed:
Merge commit 'f59a5b36db56a178d6ead446d8d444e6a0598e02' into control_llik

# Conflicts:
#	man/ergm-terms.Rd

handcock committed:
Extend the exclude code and robustify it.

handcock committed:
Added direct deterministic gradients and hessian for Kpenalty.

These required extensive algebra and computation. However, they dramatically improve the computational speed and accuracy of the Kpenalty method.

krivit commented on DESCRIPTION:
```diff
@@ -48,6 +48,7 @@ Suggests:
     tergm (>= 4.0.0),
     ergm.count (>= 4.0),
     ergm.userterms (>= 3.10.0),
+    ergm.tapered (>= 1.0.0),
```
Since ergm.tapered is not on CRAN, we also need to add a Remotes: directive.

That said, we might want to keep it away from requirements to avoid a chicken-and-egg problem the first time we release.

krivit commented on R/ergm_estfun.R:
```diff
@@ -30,36 +30,69 @@
 #' @keywords internal
 #' @export
-ergm.estfun <- function(stats, theta, model, ...){
+ergm.estfun <- function(stats, theta, model, exclude=NULL, ...){
```
I am not sure whether this is a good idea in its current form:
* It doesn't seem to do anything that subsetting the function's output wouldn't already. Both matrices and `mcmc.list` objects already support column subsetting. If it did the subsetting prior to the estimating function calculation, that might save computing time, but it doesn't do that, as far as I can tell.
* Matching columns by name can be very unreliable, because in `ergm` you *can* have duplicate coefficient names, at least at the moment.

handcock replied:
Finally replying to this. I tried the subsetting but can not get it to work as there are also places like `ergm.bridge` where it is needed. In the end I just left it as an option as it is cleaner and more flexible.

krivit commented on R/ergm.stocapprox.R:
```diff
@@ -57,6 +57,12 @@ ergm.stocapprox <- function(init, nw, model,
     target.stats <- model$target.stats
   })
+
+  control.llik <- list(MCMLE.metric=control$MCMLE.metric, MCMLE.trustregion=control$SA.trustregion,
```
Do we want to implement a `control.ergm.llik.<METRIC>()` family of functions so that the user could potentially specify different parameters for each metric and get argument completion etc.?

handcock replied:
Maybe latter. At the moment the list is small and has not changed for a long while (until `ergm.tapered`/Kpenalty. At some point it may be worth adding if/when the list grows

krivit commented on R/control.ergm.R:
```diff
@@ -542,12 +549,13 @@ control.ergm<-function(drop=TRUE,
                        MCMLE.MCMC.precision=if(startsWith("confidence", MCMLE.termination[1])) 0.1 else 0.005,
                        MCMLE.MCMC.max.ESS.frac=0.1,
                        MCMLE.metric=c("lognormal", "logtaylor",
                          "Median.Likelihood",
-                         "EF.Likelihood", "naive"),
+                         "EF.Likelihood", "Kpenalty", "naive"),
```
Point of clarification: is Kpenalty an alternative way to estimate the likelihood ratio, or is it estimating the penalised likelihood?

handcock replied:
Kpenalty estimates the penalized likelihood. Explicitly, is computed the log-likelihood ratio (using `lognormal`) and then the kurtosis estimates to add to the log-likelihood ratio as the penalty. Otherwise, it is close to the standard `lognormal` version. The gradient is computed (mostly) exactly.

handcock committed:
Merge branch 'master' into control_llik

handcock committed:
Added a MPLE.save.xmat control argument

Added a MPLE.save.xmat control argument to return the design matrix from the MPLE fit (if any). The current and retained default is to remove it.

handcock committed:
Added a MCMC.esteq.exclude.statistics control argument

control$MCMC.esteq.exclude.statistics allows network statistics to not be included in the estimating equations. There equations determine most convergence diagnostics. For some models there are some terms that are not zero at convergence and should be removed.  For example, this supports tapering models which may have terms with non-specific means. The estimating equations are used in a number of places and the code follows those places.

Specifically, it is a vector of  strings being the names of the statistics to be excluded from
the estimating equations.

handcock committed:
Added tests and fixed important typo

It now passes R CMD check

handcock committed:
Removed simplification that roils parallel
